File: python/백준/집합과 맵/Union_find/20040 사이클 게임.py
def find_parent(parent, x):
    if parent[x] != x:
        parent[x] = find_parent(parent, parent[x])
    return parent[x]
    
def union_parent(parent, a, b):
    a = find_parent(parent, a)
    b = find_parent(parent, b)
    

    if a == b:
        return True
        
    if a < b:
        parent[b] = a
    else:
        parent[a] = b
        
import sys

input = sys.stdin.readline

N, M = map(int, input().split())

parent = [0] * N
for i in range(N):
    parent[i] = i


for i in range(1, M + 1):
    a, b = map(int, input().split())
    if union_parent(parent, a, b):
        print(i)
        sys.exit(0)

print(0)

File: cpp/백준/집합과 맵/Union_find/20040 사이클 게임.cpp
#include <iostream>
#include <vector>
using namespace std;

vector<int> parent;

int find_parent(int x) {
    if (parent[x] != x) {
        parent[x] = find_parent(parent[x]);
    }
    return parent[x];
}

bool union_parent(int a, int b) {
    a = find_parent(a);
    b = find_parent(b);
    if (a == b) return true;
    if (a < b) parent[b] = a;
    else parent[a] = b;
    return false;
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);

    int N, M;
    cin >> N >> M;

    parent.assign(N, 0);
    for (int i = 0; i < N; ++i) parent[i] = i;

    for (int i = 1; i <= M; ++i) {
        int a, b;
        cin >> a >> b;
        if (union_parent(a, b)) {
            cout << i << "\n";
            return 0;
        }
    }

    cout << 0 << "\n";
    return 0;
}
